src/services/textChanges.ts

```diff
@@ -349,6 +352,23 @@ namespace ts.textChanges {
             this.replaceRange(sourceFile, { pos, end: pos }, createToken(modifier), { suffix: " " });
         }
 
+        public insertCommentBeforeLine(sourceFile: SourceFile, lineNumber: number, position: number, commentText: string): void {

```

Comment:
I'm surprised that we need to compute line numbers for this operation (I would have guessed, perhaps incorrectly, that that was expensive), but I see that this is just a refactoring.

Reply:
It could probably be done without the line numbers by counting back to the beginning of the line. Though, getting the line number is just a binary search given that `sourceFile.lineMap` is cached.